### ISSUE TITLE
Grep plugins have awful performance

### ISSUE THREAD

**Issue by andresriancho** [andresriancho/w3af#15982]
## User story
As a user I'm seeing really slow HTTP requests per second in w3af.

## Debugging and first fix
Grep plugins read the request and responses from the grep queue.

The grep queue is populated from the extended URL library. If grep plugins are fast, the queue work is consumed faster than HTTP requests are added.

Using the features added to the producer / consumer queues (see: https://github.com/andresriancho/w3af/tree/feature/smarter-queue) I was able to identify that the grep queue was quickly filled and thus the extended URL library was locked, waiting for the grep plugins to do their work.

I fixed this in the same branch, using an in-memory queue which overflows to disk. This queue never fills up and thus it allows the extended URL library to continue without locking.

## Second problem
I just applied some changes which allow us to understand how much time grep plugins take to run, and the results are awful. This shows grep plugins that took more than 1 second during the first five
minutes of a scan to target.com:

```
[Mon Aug 28 10:06:43 2017 - debug]
ssn.grep(https://www.target.com/) took 1.19 seconds to run
[Mon Aug 28 10:07:02 2017 - debug]
path_disclosure.grep(https://www.target.com/) took 12.13 seconds
to run
[Mon Aug 28 10:07:02 2017 - debug]
credit_cards.grep(https://www.target.com/) took 1.17 seconds to
run
[Mon Aug 28 10:07:02 2017 - debug]
code_disclosure.grep(https://www.target.com/) took 3.75 seconds to
run
[Mon Aug 28 10:07:19 2017 - debug]
ssn.grep(https://www.target.com/?view=phpinfo) took 1.08 seconds
to run
[Mon Aug 28 10:07:19 2017 - debug]
path_disclosure.grep(https://www.target.com/?view=phpinfo) took
11.92 seconds to run
[Mon Aug 28 10:07:26 2017 - debug]
code_disclosure.grep(https://www.target.com/?view=phpinfo) took
3.74 seconds to run
[Mon Aug 28 10:07:29 2017 - debug]
get_emails.grep(https://www.target.com/phpVersion.php) took 1.84
seconds to run
[Mon Aug 28 10:07:30 2017 - debug]
get_emails.grep(https://www.target.com/site-manifest.gears) took
1.22 seconds to run
[Mon Aug 28 10:07:44 2017 - debug]
path_disclosure.grep(https://www.target.com/?mode=phpinfo) took
11.90 seconds to run
[Mon Aug 28 10:07:46 2017 - debug]
credit_cards.grep(https://www.target.com/?mode=phpinfo) took 1.01
seconds to run
[Mon Aug 28 10:07:50 2017 - debug]
code_disclosure.grep(https://www.target.com/?mode=phpinfo) took
3.76 seconds to run
[Mon Aug 28 10:08:38 2017 - debug]
get_emails.grep(https://www.target.com/manifest.json) took 7.09
seconds to run
[Mon Aug 28 10:08:38 2017 - debug]
strange_parameters.grep(https://www.target.com/manifest.json) took
1.60 seconds to run
[Mon Aug 28 10:08:39 2017 - debug]
get_emails.grep(https://www.target.com/gears-manifest.json) took
14.68 seconds to run
[Mon Aug 28 10:08:41 2017 - debug]
get_emails.grep(https://www.target.com/_sys/ajax/module/.listing)
took 11.17 seconds to run
[Mon Aug 28 10:08:42 2017 - debug]
get_emails.grep(https://www.target.com/x.php) took 1.43 seconds to
run
[Mon Aug 28 10:08:42 2017 - debug]
get_emails.grep(https://www.target.com/site_manifest.gears) took
8.30 seconds to run
[Mon Aug 28 10:08:43 2017 - debug]
get_emails.grep(https://www.target.com/TEST.php?mode=phpinfo) took
1.43 seconds to run
[Mon Aug 28 10:08:44 2017 - debug]
get_emails.grep(https://www.target.com/PHPversion.php) took 2.98
seconds to run
[Mon Aug 28 10:08:49 2017 - debug]
get_emails.grep(https://www.target.com/gearsconfig.php) took 2.44
seconds to run
[Mon Aug 28 10:09:17 2017 - debug]
get_emails.grep(https://www.target.com/cache-manifest.php) took
6.67 seconds to run
[Mon Aug 28 10:09:18 2017 - debug]
get_emails.grep(https://www.target.com/gearsmanifest.json) took
3.39 seconds to run
[Mon Aug 28 10:09:18 2017 - debug]
get_emails.grep(https://www.target.com/phpInfo.php) took 1.83
seconds to run
[Mon Aug 28 10:09:19 2017 - debug]
get_emails.grep(https://www.target.com/xxx.php) took 1.60 seconds
to run
[Mon Aug 28 10:09:19 2017 - debug]
get_emails.grep(https://www.target.com/xx.php) took 1.03 seconds
to run
[Mon Aug 28 10:09:20 2017 - debug]
get_emails.grep(https://www.target.com/PHPINFO.php) took 3.42
seconds to run
[Mon Aug 28 10:09:21 2017 - debug]
get_emails.grep(https://www.target.com/Rfsvo) took 2.25 seconds to
run
[Mon Aug 28 10:09:21 2017 - debug]
get_emails.grep(https://www.target.com/sitemanifest.gears) took
1.90 seconds to run
[Mon Aug 28 10:09:27 2017 - debug]
get_emails.grep(https://www.target.com/_sys/ajax/module/) took
5.64 seconds to run
[Mon Aug 28 10:09:32 2017 - debug]
get_emails.grep(https://www.target.com/offline_manifest.json) took
2.76 seconds to run
[Mon Aug 28 10:09:36 2017 - debug]
get_emails.grep(https://www.target.com/admin.php?mode=phpinfo)
took 1.03 seconds to run
[Mon Aug 28 10:09:40 2017 - debug]
get_emails.grep(https://www.target.com/gears_config.gears) took
3.13 seconds to run
[Mon Aug 28 10:09:41 2017 - debug]
get_emails.grep(https://www.target.com/sitemanifest.php) took 1.32
seconds to run
[Mon Aug 28 10:09:47 2017 - debug]
get_emails.grep(https://www.target.com/_sys/ajax/module/info.php)
took 4.25 seconds to run
[Mon Aug 28 10:09:50 2017 - debug]
get_emails.grep(https://www.target.com/gears-manifest.gears) took
2.24 seconds to run
[Mon Aug 28 10:09:53 2017 - debug]
get_emails.grep(https://www.target.com/account/deposit?amount=&method=NewGames)
took 2.26 seconds to run
```

Some of those plugins take more than 10 seconds to run!

Luckily those plugins are not blocking the request / response cycle anymore. They will be run when there is free CPU.

Sadly their performance is still very poor and they will (most likely) slow down the overall scan time: HTTP requests will be sent at a rate faster than the grep plugins can process it, they will be queued on disk, and at some point when the audit / discovery plugins stop sending so many requests the grep plugins will "catch up" and consume the items from the on disk queue. Most likely the grep plugins will "catch up" at the end of the scan when no more HTTP requests are sent, thus we're going to see 5 to 10 minutes (or more depending on the scan size) at the end of the scan where the framework doesn't send
HTTP requests but the CPU is ~100% processing requests and responses with grep plugins.

Solutions? There are a few:

 * **Must-have**, easy but potentially time consuming: Find the Top5 grep plugins that consume more time and improve their performance. Run more scans and repeat until grep plugins process request  / responses at least at the same rate at which w3af sends requests. The improvements have limits. For some plugins I'll be able to reduce the run time from 10 seconds to 0.2 sec. Others will be reduced only 5%. Hopefully the improvements we can apply are good enough to achieve our goal.

 * **Nice-to-have**, hard and time consuming: Refactor the code so that grep plugins are run in a multiprocessing pool. Right now grep plugins are only using 1 CPU core. This refactoring will allow us to
run X times faster (depending on the cores) and also allows us to kill a grep plugin that takes more than N seconds to run. Killing a plugin is not something we want to do all the time, but if there is a grep
plugin that for some specific / rare scenarios is using a lot of CPU time and takes 15 seconds to run, we would want to kill it before that.

Workarounds:

 * Enable less grep plugins. Only enable the ones that have a "good mix" between run time, false positives, things that interest your clients.

**Comment by andresriancho** [comment]
Related issue: https://github.com/andresriancho/w3af/issues/16187
Performance improvement for most grep plugins: https://github.com/andresriancho/w3af/tree/feature/ahocorasick

**Comment by andresriancho** [comment]
At least in the latest scans of one of the common targets I use for measuring performance I get that `grep.get_emails` is the slowest plugin:

```
aq-ONTNF.txt:[Wed Feb 14 18:47:08 2018 - debug] get_emails.grep(uri="http://www.target.com/images/apple-icon-76x76.png?v=iDontExist") took 18.26s to run
aq-ONTNF.txt:[Wed Feb 14 18:56:39 2018 - debug] get_emails.grep(uri="http://www.target.com/create-account?utm_source=developing_apps_without_code_no_code_no_problem&utm_medium=%n%n%n%n%n%n%n%n%n%n%n%n%n%n%n%n%n%n%n%n%n%n%n%n%n") took 25.59s to run
aq-out100.txt:[Thu Feb 15 16:44:53 2018 - debug] get_emails.grep(uri="http://www.target.com/site_manifest.txt") took 10.71s to run
aq-out100.txt:[Thu Feb 15 16:49:46 2018 - debug] get_emails.grep(uri="http://www.target.com/career/.git/index") took 26.30s to run
aq-out100.txt:[Thu Feb 15 16:51:23 2018 - debug] get_emails.grep(uri="http://www.target.com/images/apple-icon-60x60.png?v=1)&waitfor%20delay%20'0:0:12'--=") took 23.07s to run
aq-out100.txt:[Thu Feb 15 16:53:08 2018 - debug] get_emails.grep(uri="http://www.target.com/images/mobile-arrow.svg?v=sleep(12)") took 11.18s to run
aq-out100.txt:[Thu Feb 15 16:54:42 2018 - debug] get_emails.grep(uri="http://www.target.com/products/?view=w3af\rvulnerable073b:%20ae5cw3af") took 10.61s to run
aq-out100.txt:[Thu Feb 15 16:59:10 2018 - debug] get_emails.grep(uri="http://www.target.com/free-trial?utm_source=VisualCodingUpdate&utm_medium=echo%20str_repeat('azxlg',5)&utm_campaign=website") took 23.36s to run
aq-out200.txt:[Thu Feb 15 17:28:33 2018 - debug] get_emails.grep(uri="http://www.target.com/gears-manifest") took 10.15s to run
aq-out200.txt:[Thu Feb 15 17:33:38 2018 - debug] get_emails.grep(uri="http://www.target.com/upl/files/TEST.php?mode=phpinfo") took 24.16s to run
aq-out200.txt:[Thu Feb 15 17:35:22 2018 - debug] get_emails.grep(uri="http://www.target.com/images/apple-icon-76x76.png?v=|ping%20-c%2013%20localhost") took 20.05s to run
aq-out200.txt:[Thu Feb 15 17:36:56 2018 - debug] get_emails.grep(uri="http://www.target.com/products/?mode=phpinfo`/bin/cat%20/etc/passwd`") took 11.93s to run
aq-out200.txt:[Thu Feb 15 17:37:22 2018 - debug] get_emails.grep(uri="http://www.target.com/products/?mode=1&waitfor%20delay%20'0:0:12'--=") took 10.21s to run
aq-out20-1.txt:[Thu Feb 15 17:28:33 2018 - debug] get_emails.grep(uri="http://www.target.com/gears-manifest") took 10.15s to run
aq-out20-1.txt:[Thu Feb 15 17:33:38 2018 - debug] get_emails.grep(uri="http://www.target.com/upl/files/TEST.php?mode=phpinfo") took 24.16s to run
aq-out20-1.txt:[Thu Feb 15 17:35:22 2018 - debug] get_emails.grep(uri="http://www.target.com/images/apple-icon-76x76.png?v=|ping%20-c%2013%20localhost") took 20.05s to run
aq-out20-1.txt:[Thu Feb 15 17:36:56 2018 - debug] get_emails.grep(uri="http://www.target.com/products/?mode=phpinfo`/bin/cat%20/etc/passwd`") took 11.93s to run
aq-out20-1.txt:[Thu Feb 15 17:37:22 2018 - debug] get_emails.grep(uri="http://www.target.com/products/?mode=1&waitfor%20delay%20'0:0:12'--=") took 10.21s to run
aq-out20-1.txt:[Thu Feb 15 17:43:08 2018 - debug] get_emails.grep(uri="http://www.target.com/free-trial?utm_source=WebDevsNowAppDevs&utm_medium=blog\r\ntype%20%SYSTEMROOT%\win.ini&utm_campaign=August2016") took 21.86s to run
aq-XvAWD.txt:[Thu Feb 15 18:06:14 2018 - debug] get_emails.grep(uri="http://www.target.com/gearsconfig") took 10.41s to run
aq-XvAWD.txt:[Thu Feb 15 18:10:50 2018 - debug] get_emails.grep(uri="http://www.target.com/images/logo.svg?v=&type%20%SYSTEMROOT%\win.ini=") took 21.82s to run
aq-XvAWD.txt:[Thu Feb 15 18:12:51 2018 - debug] get_emails.grep(uri="http://www.target.com/images/apple-icon-180x180.png?v=w3af\rvulnerable073b:%20ae5cw3af") took 22.23s to run
aq-XvAWD.txt:[Thu Feb 15 18:15:03 2018 - debug] get_emails.grep(uri="http://www.target.com/products/?view=|/bin/cat%20/etc/passwd") took 10.56s to run
aq-XvAWD.txt:[Thu Feb 15 18:15:36 2018 - debug] get_emails.grep(uri="http://www.target.com/products/?view=print%20'joidq'x5") took 12.10s to run
aq-XvAWD.txt:[Thu Feb 15 18:20:11 2018 - debug] get_emails.grep(uri="http://www.target.com/create-account?utm_source=pricing_announcement&utm_medium=`ping%20-n%2013%20localhost`") took 20.48s to run
```

**Comment by andresriancho** [comment]
Other data is shown at:

```
egrep -e ' took ..\.' ~/Dropbox/w3af/scan-logs/* | grep grep
```

**Comment by andresriancho** [comment]
This task is blocked by https://github.com/noxdafox/pebble/issues/30 . I can't use the thread pool in grep consumer because of that issue.

The code which uses threads to run grep plugins is at:
https://github.com/andresriancho/w3af/tree/feature/grep-in-threads

Once the `pebble` issue is fixed I need to run various scans to check if this thread is faster than develop.

**Comment by andresriancho** [comment]
Short term TODO:

 - [ ] Add a cache to `get_tags_by_filter`

**Comment by andresriancho** [comment]
## Running grep plugins in threads
> Quick-fix, run grep plugins in a thread pool. The grep plugins are CPU-bound, which makes us believe that running them in a thread pool will only add overhead... but my gut tells me that most of the time grep plugins are running is waiting for a document parser to be ready (that consumes no CPU on the main thread, because documents are parsed in other processes). So, running grep plugins in a thread pool with 3 or 4 workers should speed things up: 1 or 2 threads will be locked waiting for document parsers, another thread will be CPU-bound running a grep plugin that uses regular expressions for matching.

## Results
**The scan time increased** from ~24 to ~28 minutes.

I didn't have the time to analyze exactly why that happen, but it did and it is frustrating since it goes against my gut feeling.

The code for testing this is at https://github.com/andresriancho/w3af/tree/feature/grep-in-threads

**Comment by noxdafox** [comment]
Never run CPU bound jobs in multiple threads. Depending on your Python version you might encounter different degrees of performance degradation.

You can have a pretty good decomposition of the issue at the following talk:
https://www.youtube.com/watch?v=Obt-vMVdM8s

**Comment by andresriancho** [comment]
@noxdafox yeah, thanks for the input, but w3af is a really complex beast and even though I might say that grep plugins are CPU-bound in some cases they are not.

For example, grep plugins parse HTML and w3af's HTML parser is [run in a different process](https://github.com/andresriancho/w3af/blob/master/w3af/core/data/parsers/mp_document_parser.py#L73). Thus in some cases:
 * grep plugin starts
 * grep plugin wants to parse an HTML document
 * Process from the document parser pool is used to start the HTML parsing process
 * grep plugin waits for the parsing result `here we have the benefit of using threads`
 * grep plugin processes result

Each of the main steps in the process is CPU-bound: grep plugin starting, parsing html, processing html parsing results. But "in between" we have that wait for the HTML parsing result that is done on a different process that will (most likely) benefit from threads.

I ended up with html parsers in different processes because in some rare cases the parsing hang or took too long, and I wanted to be able to kill it.

**Comment by andresriancho** [comment]
I would love to be able to be able to run grep plugins in different processes, but design decisions I took years ago are hurting progress on that front now. It is completely possible to refactor those singletons... but would require some effort...
